jerry-ext/jerry-context/include/jerry-context.h

```diff
@@ -0,0 +1,43 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRYX_CONTEXT_H
+#define JERRYX_CONTEXT_H
+
+#include "jerryscript.h"
+
+#ifndef JERRYX_CONTEXT_SLOTS
+#define JERRYX_CONTEXT_SLOTS 1
+#endif /* !JERRYX_CONTEXT_SLOTS */
+
+/**
+ * Function to pass to jerry_init_with_user_context() as its second parameter
+ */
+void *jerryx_context_init (void);
+
+/**
+ * Function to pass to jerry_init_with_user_context() as its third parameter
+ */
+void jerryx_context_deinit (void *user_context_p);
+
+/**
+ * Function to request an index for per-context data
+ */
+int jerryx_context_request_slot (jerry_user_context_init_cb init_cb,
+                                 jerry_user_context_deinit_cb deinit_cb);
```

Comment:
So, for example, the following works as expected with vanilla gcc both in one file, as shown for simplicity below, or in multiple files, as indicated by the comments:
```C
#include <stdio.h>

/* module.h */
typedef void (*module_init)(void);

/* module1.c */
static void module1_init(void) {
  printf("module1_init\n");
}
module_init __module1_init __attribute__((section("__module_init"))) = module1_init;

/* module2.c */
static void module2_init(void) {
  printf("module2_init\n");
}
module_init __module2_init __attribute__((section("__module_init"))) = module2_init;

/* main.c */
extern module_init __start___module_init;
extern module_init __stop___module_init;

int main(int argc, char **argv) {
  module_init *iter;

  for (iter = &__start___module_init; iter < &__stop___module_init; ++iter) {
    (*iter)();
  }
  return 0;
}
```
Unfortunately, when building this code within jerryscript, the two symbols `__start___module_init` and `__stop___module_init` are missing - so that's what we'd need our linker script to produce.